vpr/src/base/setup_clocks.cpp

```diff
@@ -0,0 +1,207 @@
+
+#include <string>
+#include <iostream>
+#include <sstream>
+
+static MetalLayer get_metal_layer_from_name(
+        std::string metal_layer_name,
+        std::unordered_map<std::string, t_metal_layer> clock_metal_layers,
+        std::string clock_network_name);
+static void setup_clock_network_wires(const t_arch& Arch, std::vector<t_segment_inf>& segment_inf);
+static void setup_clock_connections(const t_arch& Arch);
+static std::vector<std::string> split_clock_and_switch_names(std::string names);
+
+void setup_clock_networks(const t_arch& Arch, std::vector<t_segment_inf>& segment_inf) {
+    setup_clock_network_wires(Arch, segment_inf);
+    setup_clock_connections(Arch);
+}
+
+void setup_clock_network_wires(const t_arch& Arch, std::vector<t_segment_inf>& segment_inf) {
+
+    auto& device_ctx = g_vpr_ctx.mutable_device();
+    auto& clock_networks_device = device_ctx.clock_networks;
+    auto& grid = device_ctx.grid;
+
+    auto& clock_networks_arch = Arch.clock_networks_arch;
+    auto& clock_metal_layers = Arch.clock_metal_layers;
+
+    // TODO: copied over from SetupGrid. Ensure consistency by only assiging in one place
+    t_formula_data vars;
+    vars.set_var_value("W", grid.width());
+    vars.set_var_value("H", grid.height());
+
+    for(auto clock_network_arch : clock_networks_arch) {
+        switch(clock_network_arch.type) {
+            case e_clock_type::SPINE: {
+                clock_networks_device.emplace_back(new ClockSpine());
+                ClockSpine* spine = dynamic_cast<ClockSpine*>(clock_networks_device.back().get());
+
+                spine->set_clock_name(clock_network_arch.name);
+                spine->set_num_instance(clock_network_arch.num_inst);
+                spine->set_metal_layer(get_metal_layer_from_name(
+                        clock_network_arch.metal_layer,
+                        clock_metal_layers,
+                        clock_network_arch.name));
+                spine->set_initial_wire_location(
+                        parse_formula(clock_network_arch.wire.start, vars),
+                        parse_formula(clock_network_arch.wire.end, vars),
+                        parse_formula(clock_network_arch.wire.position, vars));
+                spine->set_wire_repeat(
+                        parse_formula(clock_network_arch.repeat.x, vars),
+                        parse_formula(clock_network_arch.repeat.y, vars));
+                spine->set_drive_location(parse_formula(clock_network_arch.drive.offset, vars));
+                spine->set_drive_switch(clock_network_arch.drive.arch_switch_idx);
+                spine->set_drive_name(clock_network_arch.drive.name);
+                spine->set_tap_locations(
+                        parse_formula(clock_network_arch.tap.offset, vars),
+                        parse_formula(clock_network_arch.tap.increment, vars));
+                spine->set_tap_name(clock_network_arch.tap.name);
+
+                spine->create_segments(segment_inf);
```

Comment:
Clarify that this is actually where the architecture segments get changes.

This function seems to conflate two things:
1) Parsing the clock arch information
2) Modifying the architecture's segment information.

Reply:
Done. Commit  bb1041e098233f8d8125d89e6ee4ea6df0251af1